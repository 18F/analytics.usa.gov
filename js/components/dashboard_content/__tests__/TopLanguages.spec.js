import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import TopLanguages from "../TopLanguages";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadDailyReportJSON: jest.fn(),
}));

describe("TopLanguages", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <TopLanguages dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "language",
        query: {},
        meta: {
          name: "Browser Language",
          description:
            "90 days of visits by browser language for all sites. (>10 sessions)",
        },
        totals: {
          visits: 4188421693,
          by_language: {
            English: 3811038327,
            Spanish: 220470343,
            Chinese: 38326396,
            French: 19117303,
            Japanese: 10171020,
            Portuguese: 11105949,
            German: 12988378,
            Italian: 6472652,
            Korean: 7082855,
            Russian: 10095454,
            Turkish: 5535267,
            Dutch: 3992621,
            Indonesian: 3077390,
            Polish: 2972422,
            Arabic: 3796983,
            Swedish: 2399832,
            Vietnamese: 2266598,
            Ukrainian: 1544534,
            "Norwegian BokmÃ¥l": 1414679,
            Danish: 1425160,
            Thai: 1294188,
          },
          by_language_code: {
            "en-us": 3592299784,
            "en-gb": 130460030,
            "es-419": 72377307,
            "es-us": 67268257,
            "es-es": 39872015,
            "zh-cn": 32226679,
            "en-ca": 29103005,
            "es-mx": 21143602,
            en: 18722007,
            "en-au": 16075630,
            "en-in": 14356792,
            "fr-fr": 12075009,
            es: 10618147,
            ja: 9335163,
            "pt-br": 9117027,
            "de-de": 9087969,
            "it-it": 5573122,
            "ko-kr": 5547917,
            "ru-ru": 5110051,
            "tr-tr": 5003308,
          },
        },
        taken_at: "2024-03-11T17:07:19.627Z",
      };

      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <TopLanguages dataHrefBase="http://www.example.com/data/" />,
      );
      await waitFor(() => screen.getByText("Chinese"));
      // Wait for barchart transition animation to complete (200 ms, set in
      // js/lib/chart_helpers/barchart.js)
      await delay(300);
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <TopLanguages dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
