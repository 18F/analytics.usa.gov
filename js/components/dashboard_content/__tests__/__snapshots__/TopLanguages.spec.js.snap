// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopLanguages when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <div
    class="grid-row"
  >
    <div
      class="chart__title display-flex card:grid-col-12 mobile-lg:grid-col-fill card:flex-justify-center mobile-lg:flex-justify-start card:padding-bottom-105 mobile-lg:padding-bottom-0"
    >
      <a
        aria-label="languages-yesterday.csv"
        href="http://www.example.com/data//languages-yesterday.csv"
      >
        <svg
          aria-hidden="true"
          class="usa-icon margin-bottom-neg-05 margin-left-05"
          focusable="false"
          role="img"
        >
          <use
            xlink:href="/assets/uswds/img/sprite.svg#file_present"
          />
        </svg>
      </a>
      <a
        href="/definitions#dimension_language"
      >
        <span
          class="usa-tooltip"
        >
          <span
            aria-describedby="tooltip-550000"
            class="usa-tooltip__trigger"
            data-position="top"
            tabindex="0"
          >
            Languages
          </span>
          <span
            aria-hidden="true"
            class="usa-tooltip__body"
            id="tooltip-550000"
            role="tooltip"
          >
            The name of the language of a user's browser or device.
          </span>
        </span>
      </a>
    </div>
    <div
      class="card:grid-col-12 mobile-lg:grid-col-auto"
    >
      <div
        class="display-flex card:flex-justify-center mobile-lg:flex-justify-end"
      >
        <select
          aria-label="language chart time filter"
          class="usa-select chart__filter margin-top-0"
        >
          <option
            value="languages-yesterday"
          >
            Yesterday
          </option>
          <option
            value="languages-7-days"
          >
            7 Days
          </option>
          <option
            value="languages-30-days"
          >
            30 Days
          </option>
          <option
            value="languages-90-days"
          >
            90 Days
          </option>
          <option
            value="languages-current-year"
          >
            Current Year
          </option>
          <option
            value="languages-current-fiscal-year"
          >
            Current Fiscal Year
          </option>
          <option
            value="languages-previous-year"
          >
            Previous Year
          </option>
          <option
            value="languages-previous-fiscal-year"
          >
            Previous Fiscal Year
          </option>
        </select>
      </div>
    </div>
  </div>
  <div
    class="text--capitalize"
  >
    <figure
      id="chart_language"
    >
      <div
        class="data chart__bar-chart margin-top-2"
      >
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              English
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              91.2%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 91%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Spanish
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              5.3%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 5%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Chinese
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.9%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 1%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              French
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.5%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 0%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              German
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.3%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 0%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Portuguese
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.3%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 0%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Japanese
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.2%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 0%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Russian
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.2%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 0%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Korean
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.2%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 0%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Italian
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.2%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 0%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Other
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              0.7%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 1%;"
            />
          </div>
        </div>
      </div>
    </figure>
  </div>
</DocumentFragment>
`;

exports[`TopLanguages when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <div
    class="grid-row"
  >
    <div
      class="chart__title display-flex card:grid-col-12 mobile-lg:grid-col-fill card:flex-justify-center mobile-lg:flex-justify-start card:padding-bottom-105 mobile-lg:padding-bottom-0"
    >
      <a
        aria-label="languages-yesterday.csv"
        href="http://www.example.com/data//languages-yesterday.csv"
      >
        <svg
          aria-hidden="true"
          class="usa-icon margin-bottom-neg-05 margin-left-05"
          focusable="false"
          role="img"
        >
          <use
            xlink:href="/assets/uswds/img/sprite.svg#file_present"
          />
        </svg>
      </a>
      <a
        href="/definitions#dimension_language"
      >
        <span
          class="usa-tooltip"
        >
          <span
            aria-describedby="tooltip-550000"
            class="usa-tooltip__trigger"
            data-position="top"
            tabindex="0"
          >
            Languages
          </span>
          <span
            aria-hidden="true"
            class="usa-tooltip__body"
            id="tooltip-550000"
            role="tooltip"
          >
            The name of the language of a user's browser or device.
          </span>
        </span>
      </a>
    </div>
    <div
      class="card:grid-col-12 mobile-lg:grid-col-auto"
    >
      <div
        class="display-flex card:flex-justify-center mobile-lg:flex-justify-end"
      >
        <select
          aria-label="language chart time filter"
          class="usa-select chart__filter margin-top-0"
        >
          <option
            value="languages-yesterday"
          >
            Yesterday
          </option>
          <option
            value="languages-7-days"
          >
            7 Days
          </option>
          <option
            value="languages-30-days"
          >
            30 Days
          </option>
          <option
            value="languages-90-days"
          >
            90 Days
          </option>
          <option
            value="languages-current-year"
          >
            Current Year
          </option>
          <option
            value="languages-current-fiscal-year"
          >
            Current Fiscal Year
          </option>
          <option
            value="languages-previous-year"
          >
            Previous Year
          </option>
          <option
            value="languages-previous-fiscal-year"
          >
            Previous Fiscal Year
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TopLanguages when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <div
    class="grid-row"
  >
    <div
      class="chart__title display-flex card:grid-col-12 mobile-lg:grid-col-fill card:flex-justify-center mobile-lg:flex-justify-start card:padding-bottom-105 mobile-lg:padding-bottom-0"
    >
      <a
        aria-label="languages-yesterday.csv"
        href="http://www.example.com/data//languages-yesterday.csv"
      >
        <svg
          aria-hidden="true"
          class="usa-icon margin-bottom-neg-05 margin-left-05"
          focusable="false"
          role="img"
        >
          <use
            xlink:href="/assets/uswds/img/sprite.svg#file_present"
          />
        </svg>
      </a>
      <a
        href="/definitions#dimension_language"
      >
        <span
          class="usa-tooltip"
        >
          <span
            aria-describedby="tooltip-550000"
            class="usa-tooltip__trigger"
            data-position="top"
            tabindex="0"
          >
            Languages
          </span>
          <span
            aria-hidden="true"
            class="usa-tooltip__body"
            id="tooltip-550000"
            role="tooltip"
          >
            The name of the language of a user's browser or device.
          </span>
        </span>
      </a>
    </div>
    <div
      class="card:grid-col-12 mobile-lg:grid-col-auto"
    >
      <div
        class="display-flex card:flex-justify-center mobile-lg:flex-justify-end"
      >
        <select
          aria-label="language chart time filter"
          class="usa-select chart__filter margin-top-0"
        >
          <option
            value="languages-yesterday"
          >
            Yesterday
          </option>
          <option
            value="languages-7-days"
          >
            7 Days
          </option>
          <option
            value="languages-30-days"
          >
            30 Days
          </option>
          <option
            value="languages-90-days"
          >
            90 Days
          </option>
          <option
            value="languages-current-year"
          >
            Current Year
          </option>
          <option
            value="languages-current-fiscal-year"
          >
            Current Fiscal Year
          </option>
          <option
            value="languages-previous-year"
          >
            Previous Year
          </option>
          <option
            value="languages-previous-fiscal-year"
          >
            Previous Fiscal Year
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
